temporal-server/authorizer/claimer.go

```diff
@@ -0,0 +1,245 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package authorizer
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/canonical/ofga"
+	"go.temporal.io/server/common/authorization"
+	"go.uber.org/zap"
+)
+
+//go:generate mockgen -destination=mocks/groups_provider_gen.go -package=mock github.com/canonical/charmed-temporal-image/temporal-server/authorizer NamespaceAccessProvider,TokenVerifier
+
+type NamespaceAccess struct {
+	Namespace string
+	Relation  string
+}
+
+// NamespaceAccessProvider is an interface that defines the method to retrieve namespace
+// access information for a given email.
+type NamespaceAccessProvider interface {
+	GetNamespaceAccessInformation(ctx context.Context, email string) ([]NamespaceAccess, error)
+}
+
+// AuthClient implements the necessary methods needed to fetch namespace access
+// information from an OpenFGA store.
+type AuthClient struct {
+	OfgaClient *ofga.Client
+}
+
+// TokenClaimMapper implements Temporal authorization.ClaimMapper,
+// performing a OIDC-based authorization.
+type TokenClaimMapper struct {
+	// TokenVerifier is used to verify the provided access token.
+	TokenVerifier TokenVerifier
+	// NamespaceAccessProvider is used to identify the namespaces that the user
+	// logging in via the access token has access to.
+	NamespaceAccessProvider NamespaceAccessProvider
```

Comment:
I guess you could have the ofga client directly here. This "class" seems to be already too ofga specific (the two authclient methods could be also part of tokenclaimapper, which could be called OfgaClaimMapper).

Reply:
I tried implementing it in a different way but it made testing slightly more difficult, so I kept the class as is, following in the same footsteps of the previous temporal k8s implementation.